### PR TITLE
make sure weights always sum up to 1, fix #4369

### LINKED ISSUE #4369: SMC-ABC update_weights_beta unstable with small epsilon
## Description of your problem

When using SMC-ABC (code below) and a small epsilon in the `Simulator` the algorithm that sets `SMC.weights` (`SMC.update_weights_beta`) becomes unstable.

Using the interactive debugger after the traceback pasted below:
```
ipdb> self.weights.sum()
1.0000004109817149
```

I have opened a PR with a suggested solution https://github.com/pymc-devs/pymc3/pull/4370.

**Please provide a minimal, self-contained, and reproducible example.**
```python
import numpy as np
import pymc3 as pm

def f(a, b):
    x_vec = np.linspace(-2, 2, 100)
    return a + b * x_vec


x_vec = np.linspace(-2, 2, 100)
np.random.seed(0)
y = f(-1, 2) + 1e-9 * np.random.normal(size=len(x_vec))

with pm.Model() as model:
    a = pm.Uniform("a", -5, 5)
    b = pm.Uniform("b", -5, 5)
    pm.Simulator(
        "simulation",
        f,
        params=(a, b),
        epsilon=1e-09,
        observed=y,
    )
    trace, sim_data = pm.sample_smc(
        chains=5,
        draws=200,
        kernel="ABC",
        parallel=True,
        save_sim_data=True,
    )
```

**Please provide the full traceback.**
```python
---------------------------------------------------------------------------
RemoteTraceback                           Traceback (most recent call last)
RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/gscratch/home/a-banijh/miniconda3/envs/split_cv_mod/lib/python3.8/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/gscratch/home/a-banijh/miniconda3/envs/split_cv_mod/lib/python3.8/multiprocessing/pool.py", line 51, in starmapstar
    return list(itertools.starmap(args[0], args[1]))
  File "/gscratch/home/a-banijh/miniconda3/envs/split_cv_mod/lib/python3.8/site-packages/pymc3/smc/sample_smc.py", line 270, in sample_smc_int
    smc.resample()
  File "/gscratch/home/a-banijh/miniconda3/envs/split_cv_mod/lib/python3.8/site-packages/pymc3/smc/smc.py", line 183, in resample
    resampling_indexes = np.random.choice(
  File "mtrand.pyx", line 932, in numpy.random.mtrand.RandomState.choice
ValueError: probabilities do not sum to 1
"""

The above exception was the direct cause of the following exception:

ValueError                                Traceback (most recent call last)
<ipython-input-43-6c41818b81f9> in <module>
     21         observed=y,
     22     )
---> 23     trace, sim_data = pm.sample_smc(
     24         chains=5,
     25         draws=200,

~/miniconda3/envs/split_cv_mod/lib/python3.8/site-packages/pymc3/smc/sample_smc.py in sample_smc(draws, kernel, n_steps, start, tune_steps, p_acc_rate, threshold, save_sim_data, save_log_pseudolikelihood, model, random_seed, parallel, chains, cores)
    189         loggers = [_log] + [None] * (chains - 1)
    190         pool = mp.Pool(cores)
--> 191         results = pool.starmap(
    192             sample_smc_int, [(*params, random_seed[i], i, loggers[i]) for i in range(chains)]
    193         )

~/miniconda3/envs/split_cv_mod/lib/python3.8/multiprocessing/pool.py in starmap(self, func, iterable, chunksize)
    370         `func` and (a, b) becomes func(a, b).
    371         '''
--> 372         return self._map_async(func, iterable, starmapstar, chunksize).get()
    373 
    374     def starmap_async(self, func, iterable, chunksize=None, callback=None,

~/miniconda3/envs/split_cv_mod/lib/python3.8/multiprocessing/pool.py in get(self, timeout)
    769             return self._value
    770         else:
--> 771             raise self._value
    772 
    773     def _set(self, i, obj):

~/miniconda3/envs/split_cv_mod/lib/python3.8/multiprocessing/pool.py in worker()
    123         job, i, func, args, kwds = task
    124         try:
--> 125             result = (True, func(*args, **kwds))
    126         except Exception as e:
    127             if wrap_exception and func is not _helper_reraises_exception:

~/miniconda3/envs/split_cv_mod/lib/python3.8/multiprocessing/pool.py in starmapstar()
     49 
     50 def starmapstar(args):
---> 51     return list(itertools.starmap(args[0], args[1]))
     52 
     53 #

~/miniconda3/envs/split_cv_mod/lib/python3.8/site-packages/pymc3/smc/sample_smc.py in sample_smc_int()
    268             _log.info(f"Stage: {stage:3d} Beta: {smc.beta:.3f}")
    269         smc.update_proposal()
--> 270         smc.resample()
    271         smc.mutate()
    272         smc.tune()

~/miniconda3/envs/split_cv_mod/lib/python3.8/site-packages/pymc3/smc/smc.py in resample()
    181     def resample(self):
    182         """Resample particles based on importance weights."""
--> 183         resampling_indexes = np.random.choice(
    184             np.arange(self.draws), size=self.draws, p=self.weights
    185         )

mtrand.pyx in numpy.random.mtrand.RandomState.choice()

ValueError: probabilities do not sum to 1

```

**Please provide any additional information below.**

## Versions and main components

* PyMC3 Version: '3.10.0'
* Theano Version: '1.0.12'
* Python Version: 3.8.6
* Operating system: CentOS
* How did you install PyMC3: conda

Tagging @aloctavodia because he wrote most of the code.

### PR DESCRIPTION
**Pull request by basnijholt**
**Thank your for opening a PR!**

Depending on what your PR does, here are a few things you might want to address in the description:
+ [ ] what are the (breaking) changes that this PR makes?
+ [ ] important background, or details about the implementation
+ [ ] are the changes—especially new features—covered by tests and docstrings?
+ [ ] consider adding/updating relevant example notebooks
+ [ ] right before it's ready to merge, mention the PR in the RELEASE-NOTES.md

This should fix #4369. However, this problem might require a different solution where we never need to do a correction at all.

### CODE REVIEW

**Comment by twiecki** [comment]
Thanks - definitely needs a test.

**Comment by michaelosthege** [comment]
This is a small and useful change. I agree with Thomas, that a test would be good.
But given this line `if np.abs(weight_err) > 1e-13:` I could understand if it is difficult to come up with a minimal reproducible example. @basnijholt do you have one by any chance?

File: pymc3/smc/smc.py [hunk 1]
```diff
@@ -177,6 +177,11 @@ def update_weights_beta(self):
         self.log_marginal_likelihood += logsumexp(log_weights_un) - np.log(self.draws)
         self.beta = new_beta
         self.weights = np.exp(log_weights)
+        weight_err = 1 - self.weights.sum()
```
Comment (aloctavodia) [inline pymc3/smc/smc.py]: I think doing `weights /= weights.sum()` should work

**Comment by aloctavodia** [comment]
Hi @basnijholt thanks for your contribution and sorry for the delay in answering.

**Review by aloctavodia — APPROVED** [review]

**Review by michaelosthege — APPROVED** [review]

**Pull request merged**